README.md
# mahjong_score

It is a package that calculates the score of Mahjong.
Scores are calculated based on "hu" and "fan".
It should be noted that this package does not distribute points to each player, so you need to implement it yourself.

## Usage

### BasePoint class
‘BasePoint’ is a class that handles the base points of points calculated by ‘hu’ and ‘fan’.

```dart
import 'package:mahjong_score/score.dart';

final basePoint = BasePoint(hu: 30, fan: 4);
```
|Argument|description|
|---|---|
|hu|符点数　int|
|fan|翻点数　int|

### FixedPoint class
‘FixedPoint’ deals with base points such as ‘Mangan’ and ‘Yakuman’ that do not calculate ‘hu’ and ‘fan’.
You can get an instance from the FixedPointType defined in the Enum.

```dart
import 'package:mahjong_score/score.dart';

final fixed = FixedPointType.value;
final mangan = FixedPointType.Mangan.detail;
```

### Score class
The ‘Score’ class is from the calculated base point
We will handle the points to be finally handed over in consideration of the way the game ended.

```dart
import 'package:mahjong_score/score.dart';

final score = Score(
  abstractPoint: basePoint,
  isHost: false,
  isPicked: false,
  consecutivelyConut: 0,
);
```

|Argument|description|
|---|---|
|abstractPoint|Class that inherits abstractPoint|
|isHost|Whether the winner is a parent|
|isPicked|The winner end with'Tsumo'|
|consecutivelyCount|Number of consecutive Hosts|

CHANGELOG.md
## [0.0.5] - 2021/08/22

* Supports null safety
* Update Pubspec version

## [0.0.4] - 2021/08/22

* Update Change Logs
* Update Pubspec version

## [0.0.3] - 2021/08/22

* Change Score Parameter Type
* Update Version

## [0.0.2] - 2021/02/25

* Change to README.md
* Rename 'noMoreReaderCount' -> 'consecutivelyCount'
* Add example script

## [0.0.1] - 2021/02/21

* Initial Open Source release.


lib/base.dart
import 'dart:math';

import 'abstract.dart';
import 'fixed.dart';

/// 基底ポイント
///
/// 符、翻から求めた点計算の基底となるポイント
class BasePoint extends AbstractPoint {
  /// 符
  final int hu;

  /// 翻
  final int fan;

  BasePoint({required this.hu, required this.fan}) : super();

  @override
  int get point {
    var _hu = hu;

    // 平和、七対子の符は切り上げを無視
    if (_hu != 20 && _hu != 25) {
      _hu = (_hu / 10).ceil() * 10;
    }

    // メインの点数計算式
    int _point = (_hu * 4 * pow(2, fan)).toInt();

    // 満貫になる点数は満貫の固定値点数を与える
    if (_point >= 2240) {
      _point = FixedPointType.Mangan.detail.point;
    }

    // 翻で固定値点数になる場合は固定値点数に置き換える
    if (6 <= fan && fan <= 7) {
      _point = FixedPointType.Haneman.detail.point;
    } else if (8 <= fan && fan <= 10) {
      _point = FixedPointType.Baiman.detail.point;
    } else if (11 <= fan && fan <= 12) {
      _point = FixedPointType.Sanbaiman.detail.point;
    } else if (13 <= fan) {
      _point = FixedPointType.Yakuman.detail.point;
    }
    return _point;
  }
}


lib/fixed.dart
import 'abstract.dart';

/// 固定ポイント
///
/// 一定の基準を満たし、
/// 符、翻の計算から外れた固定的なポイント
class FixedPoint extends AbstractPoint {
  @override
  final int point;

  FixedPoint({required String name, required this.point}) : super(name: name);
}

/// 固定値点数
enum FixedPointType {
  /// 満貫
  Mangan,

  /// 跳満
  Haneman,

  /// 倍満
  Baiman,

  /// 三倍満
  Sanbaiman,

  /// 役満
  Yakuman,
}

/// 満貫以上表示点数計算用列列挙拡張
extension FixedPointExtension on FixedPointType {
  static final details = <FixedPointType, FixedPoint>{
    FixedPointType.Mangan: FixedPoint(name: '満貫', point: 2000),
    FixedPointType.Haneman: FixedPoint(name: '跳満', point: 3000),
    FixedPointType.Baiman: FixedPoint(name: '倍満', point: 4000),
    FixedPointType.Sanbaiman: FixedPoint(name: '三倍満', point: 6000),
    FixedPointType.Yakuman: FixedPoint(name: '役満', point: 8000),
  };

  /// 点数
  FixedPoint get detail => details[this]!;
}


lib/pages.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'base.dart';
import 'fixed.dart';
import 'score.dart';

class PointSelector extends StatelessWidget {
  final bool isHost;
  final bool isPicked;
  final int noMoreReaderCount;
  final List<int> hus = [20, 25, 30, 40, 50, 60, 70];
  final List<int> hons = [1, 2, 3, 4];
  final List<int> specialHans = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

  PointSelector({
    this.isHost,
    this.isPicked,
    this.noMoreReaderCount,
  });

  String get playerGrade => isHost ? '親' : '子';

  String get finishMethod => isPicked ? 'ツモ' : 'ロン';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Point')),
      body: Column(
        children: [
          Text('$playerGrade  $finishMethod'),
          MahjongPointTable(
            isHost: isHost,
            isPicked: isPicked,
            noMoreReaderCount: noMoreReaderCount,
          ),
          Expanded(
            child: ListView.builder(
              itemCount: FixedPointType.values.length,
              itemBuilder: (context, i) {
                final score = Score(
                  abstractPoint: FixedPointType.values[i].detail,
                  isHost: isHost,
                  isPicked: isPicked,
                  noMoreReaderCount: noMoreReaderCount,
                );
                return ListTile(
                  title: Text(score.toString()),
                  onTap: () => Navigator.pop(context, score),
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class MahjongPointTable extends StatelessWidget {
  final bool isHost;
  final bool isPicked;
  final int noMoreReaderCount;
  final List<int> hus = [20, 25, 30, 40, 50, 60, 70];
  final List<int> fans = [1, 2, 3, 4];
  final List<int> specialHans = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

  MahjongPointTable({
    this.isHost,
    this.isPicked,
    this.noMoreReaderCount,
  });

  @override
  Widget build(BuildContext context) {
    return Table(
      border: TableBorder.all(),
      children: [
        TableRow(
          children: [
            _buildHeaderCell('符\\翻'),
            ...List.generate(
              fans.length,
              (i) => _buildHeaderCell('${fans[i].toString()}'),
            ),
          ],
        ),
        ...List.generate(hus.length, (index) {
          final hu = hus[index];
          return TableRow(
            children: [
              _buildBodyCell('$hu'),
              ...List.generate(fans.length, (index) {
                final score = Score(
                  abstractPoint: BasePoint(hu: hu, fan: fans[index]),
                  isHost: isHost,
                  isPicked: isPicked,
                  noMoreReaderCount: noMoreReaderCount,
                );
                if (score.isNotDisplay) {
                  return Container();
                }
                return GestureDetector(
                  child: _buildBodyCell(score.toString()),
                  onTap: () => Navigator.pop(context, score),
                );
              }),
            ],
          );
        }),
      ],
    );
  }

  Widget _buildHeaderCell(String label) {
    return Container(
      padding: EdgeInsets.only(top: 2.0, bottom: 2.0),
      child: Center(child: Text(label, textAlign: TextAlign.center)),
    );
  }

  Widget _buildBodyCell(String label) {
    return Container(
      padding: EdgeInsets.only(top: 10.0, bottom: 10.0),
      child: Center(child: Text(label, textAlign: TextAlign.center)),
    );
  }
}


lib/score.dart
import 'abstract.dart';

class Score {
  final AbstractPoint basePoint;

  /// 勝者が東家か
  final bool? isHost;

  /// 上がり方がツモ
  final bool? isPicked;

  /// 連荘している回数
  final int? consecutivelyCount;

  Score({
    required this.basePoint,
    this.isHost,
    this.isPicked,
    this.consecutivelyCount,
  });

  bool? get isUniform => isHost;

  /// 表示を行うか
  ///
  /// 基底ポイントが200以下の場合は表示しない
  bool get isNotDisplay => basePoint.point <= 200;

  /// ベースとなる点の子と親別の点数
  num get _point => basePoint * (isHost! ? 1.5 : 1);

  /// 合計点数(切り上げ前)
  int get sumPoint => _point * 4 as int;

  /// 合計点数(ロンのときの点数)
  int get point => _ceil(sumPoint);

  /// 親の受け取れる点数
  int get hostPoint => payOtherPoint * 3 + (300 * consecutivelyCount!);

  /// 親の支払い点数(ツモのときの点数)
  int get payHostPoint => _ceil(basePoint * 2) + (100 * consecutivelyCount!);

  /// 子の受け取れる点数
  int get otherPoint {
    return payOtherPoint * 2 + payHostPoint + (300 * consecutivelyCount!);
  }

  /// 子の支払い点数(ツモのときの点数)
  int get payOtherPoint {
    return _ceil(isHost! ? sumPoint / 3 : _point) + (100 * consecutivelyCount!);
  }

  /// 100点以下の点数はすべて切り上げする
  int _ceil(num number) => (number / 100).ceil() * 100;

  /// 表示用
  String toString() {
    var text = point.toString();
    if (basePoint.name != null) {
      text += ' (${basePoint.name})';
    }
    text += '\n${payOtherPoint.toString()}';
    text += isUniform! ? ' All' : '/${payHostPoint.toString()}';
    return text;
  }
}


example/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:mahjong_score/base.dart';
import 'package:mahjong_score/fixed.dart';
import 'package:mahjong_score/score.dart';

class PointSelector extends StatelessWidget {
  final bool? isHost;
  final bool? isPicked;
  final int? consecutivelyCount;
  final List<int> hus = [20, 25, 30, 40, 50, 60, 70];
  final List<int> hons = [1, 2, 3, 4];
  final List<int> specialHans = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

  PointSelector({
    this.isHost,
    this.isPicked,
    this.consecutivelyCount,
  });

  String get playerGrade => isHost! ? '親' : '子';

  String get finishMethod => isPicked! ? 'ツモ' : 'ロン';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('Point')),
      body: Column(
        children: [
          Text('$playerGrade  $finishMethod'),
          MahjongPointTable(
            isHost: isHost,
            isPicked: isPicked,
            noMoreReaderCount: consecutivelyCount,
          ),
          Expanded(
            child: ListView.builder(
              itemCount: FixedPointType.values.length,
              itemBuilder: (context, i) {
                final score = Score(
                  basePoint: FixedPointType.values[i].detail,
                  isHost: isHost,
                  isPicked: isPicked,
                  consecutivelyCount: consecutivelyCount,
                );
                return ListTile(
                  title: Text(score.toString()),
                  onTap: () => Navigator.pop(context, score),
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

class MahjongPointTable extends StatelessWidget {
  final bool? isHost;
  final bool? isPicked;
  final int? noMoreReaderCount;
  final List<int> hus = [20, 25, 30, 40, 50, 60, 70];
  final List<int> fans = [1, 2, 3, 4];
  final List<int> specialHans = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14];

  MahjongPointTable({
    this.isHost,
    this.isPicked,
    this.noMoreReaderCount,
  });

  @override
  Widget build(BuildContext context) {
    return Table(
      border: TableBorder.all(),
      children: [
        TableRow(
          children: [
            _buildHeaderCell('符\\翻'),
            ...List.generate(
              fans.length,
              (i) => _buildHeaderCell('${fans[i].toString()}'),
            ),
          ],
        ),
        ...List.generate(hus.length, (index) {
          final hu = hus[index];
          return TableRow(
            children: [
              _buildBodyCell('$hu'),
              ...List.generate(fans.length, (index) {
                final score = Score(
                  basePoint: BasePoint(hu: hu, fan: fans[index]),
                  isHost: isHost,
                  isPicked: isPicked,
                  consecutivelyCount: noMoreReaderCount,
                );
                if (score.isNotDisplay) {
                  return Container();
                }
                return GestureDetector(
                  child: _buildBodyCell(score.toString()),
                  onTap: () => Navigator.pop(context, score),
                );
              }),
            ],
          );
        }),
      ],
    );
  }

  Widget _buildHeaderCell(String label) {
    return Container(
      padding: EdgeInsets.only(top: 2.0, bottom: 2.0),
      child: Center(child: Text(label, textAlign: TextAlign.center)),
    );
  }

  Widget _buildBodyCell(String label) {
    return Container(
      padding: EdgeInsets.only(top: 10.0, bottom: 10.0),
      child: Center(child: Text(label, textAlign: TextAlign.center)),
    );
  }
}
